File: Data Structures & Algorithms/binary-tree-postorder-traversal/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        curr = root
        while curr:
            if not curr.right:
                res.insert(0, curr.val)
                curr = curr.left
            else:
                prev = curr.right
                while prev.left and prev.left != curr:
                    prev = prev.left
                if not prev.left:
                    res.insert(0, curr.val)
                    prev.left = curr
                    curr = curr.right
                else:
                    prev.left = None
                    curr = curr.left
        return res

File: Data Structures & Algorithms/delete-leaves-with-a-given-value/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode removeLeafNodes(TreeNode root, int target) {
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode lastVisited = null;
        TreeNode curr = root;
        stack.push(curr);
        while (!stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }

            var top = stack.peek();
            if (top.right != null && top.right != lastVisited) {
                curr = top.right;
            } else {
                lastVisited = top;
                stack.pop();
                if (top.val == target && top.left == null && top.right == null) {
                    TreeNode parent = stack.peekFirst();
                    if (parent == null) {
                        return null;
                    } else {
                        if (parent.left == top) {
                            parent.left = null;
                        } else {
                            parent.right = null;
                        }
                    }
                }
            }
        }
        return root;
    }
}